lib/backend/registrybackend/security/security.go

```diff
@@ -36,79 +50,155 @@ type Config struct {
+				Actions:    []string{"pull", "push"},
+			},
+		},
+		ClientID: "docker",
+	}))
+	return transport.NewTransport(a.roundTripper, auth.NewAuthorizer(a.challengeManager, basicHandler, bearerHandler.(auth.AuthenticationHandler)))
+}
 
-	shouldUseBasicAuth := (c.BasicAuth != nil || c.RemoteCredentialsStore != "")
-	tlsClientConfig, err := c.TLS.BuildClient()
+func (a *authenticator) updateChallenge() error {
+	resp, err := httputil.Send(
+		"GET",
+		fmt.Sprintf(basePingQuery, a.address),
+		httputil.SendTLSTransport(a.roundTripper),
+		httputil.SendAcceptedCodes(http.StatusOK, http.StatusUnauthorized),
+	)
 	if err != nil {
-		return nil, fmt.Errorf("build tls config: %s", err)
+		return err
 	}
-	if !shouldUseBasicAuth {
-		return httputil.SendTLS(tlsClientConfig), nil
+	versions := auth.APIVersions(resp, registryVersionHeader)
+	for _, version := range versions {
+		if version == v2Version {
+			if err := a.challengeManager.AddResponse(resp); err != nil {
+				return fmt.Errorf("add response: %s", err)
+			}
+			return nil
+		}
 	}
+	return fmt.Errorf("registry is not v2")
+}
 
-	authConfig, err := c.getCredentials(c.RemoteCredentialsStore, addr)
-	if err != nil {
-		return nil, fmt.Errorf("get credentials: %s", err)
+type credentialStore struct {
+	address string
+	config  Config
+}
+
+func newCredentialStore(address string, config Config) *credentialStore {
+	return &credentialStore{
+		address: address,
+		config:  config,
 	}
-	tr := http.DefaultTransport.(*http.Transport)
-	tr.TLSClientConfig = tlsClientConfig // If tlsClientConfig is nil, default is used.
-	rt, err := BasicAuthTransport(addr, repo, tr, authConfig)
-	if err != nil {
-		return nil, fmt.Errorf("basic auth: %s", err)
+}
+
+func (c credentialStore) Basic(*url.URL) (string, string) {
+	if username, password := c.credentialsFromHelper(); username != "" && username != tokenUsername {
+		return username, password
+	}
+	basic := c.config.BasicAuth
+	if basic == nil {
+		return "", ""
 	}
-	return httputil.SendTLSTransport(rt), nil
+	return basic.Username, basic.Password
 }
 
-func (c Config) getCredentials(helper, addr string) (types.AuthConfig, error) {
-	var authConfig types.AuthConfig
-	if c.BasicAuth != nil {
-		authConfig = *c.BasicAuth
+func (c credentialStore) RefreshToken(*url.URL, string) string {
+	if username, token := c.credentialsFromHelper(); username == tokenUsername {
+		return token
 	}
-	var err error
-	if helper != "" {
-		authConfig, err = c.getCredentialFromHelper(helper, addr)
-		if err != nil {
-			return types.AuthConfig{}, fmt.Errorf("get credentials from helper %s: %s", helper, err)
-		}
+	basic := c.config.BasicAuth
+	if basic == nil {
+		return ""
 	}
-	return authConfig, nil
+	return basic.IdentityToken
 }
 
-func (c Config) getCredentialFromHelper(helper, addr string) (types.AuthConfig, error) {
-	switch helper {
+func (c credentialStore) credentialsFromHelper() (string, string) {
+	switch c.config.RemoteCredentialsStore {
+	case "":
```

Comment:
im assuming this is for public repositories on dockerhub. Could you add a comment?